### PR TITLE
Controller Logic for Quick Tests

### PR DESCRIPTION
**Pull request by f11h**

### CODE REVIEW

**Review by ascheibal — DISMISSED** [review]

File: src/main/java/app/coronawarn/quicktest/controller/QuickTestCreationController.java [hunk 1]
```diff
@@ -45,41 +54,105 @@
...
+              utilities.getIdsFromToken(),
+                    shortHash, quickTestUpdateRequest.getResult());
+        } catch (QuickTestServiceException e) {
+            if (e.getReason() == QuickTestServiceException.Reason.UPDATE_NOT_FOUND) {
+                return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+            } else {
+                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+            }
+        }
+        return ResponseEntity.status(204).build();
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/controller/QuickTestCreationController.java:123]:
```suggestion
        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```

File: src/main/java/app/coronawarn/quicktest/controller/QuickTestCreationController.java [hunk 2]
```diff
@@ -45,41 +54,105 @@
...
+              utilities.getIdsFromToken(),shortHash,
+                    modelMapper.map(quickTestPersonalDataRequest, QuickTest.class));
+        } catch (QuickTestServiceException e) {
+            if (e.getReason() == QuickTestServiceException.Reason.UPDATE_NOT_FOUND) {
+                return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+            } else {
+                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+            }
+        }
+        return ResponseEntity.status(204).build();
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/controller/QuickTestCreationController.java:155]:
```suggestion
        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
```

File: src/main/java/app/coronawarn/quicktest/model/QuickTestPersonalDataRequest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    @NotNull
+    private String testBrandId;
+
+    private String testBrandName;
+
+    @ValidName
+    @NotNull
+    private String lastName;
+    @ValidName
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/model/QuickTestPersonalDataRequest.java:30]:
```suggestion
    private String lastName;
    
    @ValidName
```

File: src/main/java/app/coronawarn/quicktest/model/QuickTestPersonalDataRequest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    @NotNull
+    private String houseNumber;
+
+    // TODO
+    @NotNull
+    private String zipCode;
+
+    // TODO
+    @NotNull
+    private String city;
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/model/QuickTestPersonalDataRequest.java:59]: why is this still TODO?

File: src/main/java/app/coronawarn/quicktest/model/TestResult.java [hunk 5]
```diff
@@ -21,18 +21,52 @@
...
 public class TestResult {
 
-    private final int testResult;
+    /**
+     * Hash (SHA256) of test result id (aka QR-Code, GUID) encoded as hex string.
+     */
+    @NotBlank
+    @Pattern(regexp = "^([A-Fa-f0-9]){64}$")
+    private String id;
+    /**
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/model/TestResult.java:50]:
```suggestion

        /**
```

File: src/main/java/app/coronawarn/quicktest/service/QuickTestService.java [hunk 6]
```diff
@@ -22,34 +22,143 @@
...
 
     private final QuickTestRepository quickTestRepository;
-
     private final TestResultService testResultService;
 
-    public void saveQuickTest(QuickTest quickTest) {
+    @Value("${quick-test.tenantIdKey}")
+    String tenantIdKey;
+    @Value("${quick-test.tenantSpotIdKey}")
+    String tenantSpotIdKey;
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/service/QuickTestService.java:47]: Please replace with Spring Configuration Properties (like TestResultServerValuesConfig)

File: src/main/java/app/coronawarn/quicktest/utils/Utilities.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+@Slf4j
+@Service
+public class Utilities {
+
+    @Value("${quick-test.tenantIdKey}")
+    String tenantIdKey;
+    @Value("${quick-test.tenantSpotIdKey}")
+    String tenantSpotIdKey;
+    @Value("${quick-test.testSpotIdName}")
+    String testSpotIdName;
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/utils/Utilities.java:24]: Same as above - please as ConfigurationProperties

File: src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ */
+@Target({ElementType.METHOD, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = GuidValidator.class)
+public @interface ValidGuid {
+    /**
+     * message.
+     * @return message
+     */
+    String message() default "{Invalid guid}";
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java:22]: ?

File: src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * groups.
+     * @return groups
+     */
+    Class<?>[] groups() default {};
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java:28]: ?

File: src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * payload.
+     * @return paylaod
+     */
+    Class<? extends Payload>[] payload() default {};
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidGuid.java:34]: ?

File: src/main/java/app/coronawarn/quicktest/validation/ValidName.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+     * groups.
+     * @return groups
+     */
+    Class<?>[] groups() default {};
+
+    /**
+     * payload.
+     * @return paylaod
+     */
+    Class<? extends Payload>[] payload() default {};
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidName.java:35]: ?

File: src/main/java/app/coronawarn/quicktest/validation/ValidPhoneNumber.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+     * groups.
+     * @return groups
+     */
+    Class<?>[] groups() default {};
+
+    /**
+     * payload.
+     * @return paylaod
+     */
+    Class<? extends Payload>[] payload() default {};
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidPhoneNumber.java:34]: ?

File: src/main/java/app/coronawarn/quicktest/validation/ValidShortHash.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+     * @return groups
+    */
+    Class<?>[] groups() default {};
+
+    /**
+     * payload.
+     * @return paylaod
+     */
+    Class<? extends Payload>[] payload() default {};
+}
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/ValidShortHash.java:35]: ?

File: src/main/java/app/coronawarn/quicktest/validation/validators/GuidValidator.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package app.coronawarn.quicktest.validation.validators;
+
+import app.coronawarn.quicktest.validation.ValidGuid;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import org.apache.commons.lang3.StringUtils;
+
+public class GuidValidator  implements ConstraintValidator<ValidGuid, String> {
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/validators/GuidValidator.java:10]:
```suggestion
public class GuidValidator implements ConstraintValidator<ValidGuid, String> {
```

File: src/main/java/app/coronawarn/quicktest/validation/validators/GuidValidator.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    private static final String GUID_PATTERN = "^([A-Fa-f0-9]{2}){32}$";
+
+    @Override
+    public void initialize(ValidGuid constraintAnnotation) {
+        // nothing to initialize
+    }
+
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/validators/GuidValidator.java:17]:
```suggestion
```

Is this necessary?

File: src/main/java/app/coronawarn/quicktest/validation/validators/NameValidator.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+public class NameValidator implements ConstraintValidator<ValidName, String> {
+    private static final String NAME_PATTERN = "^[^- '](?=(?![A-Z]?[A-Z]))(?=(?![a-z]+[A-Z]))"
+            + "(?=(?!.*[A-Z][A-Z]))(?=(?!.*[- '][- '.]))(?=(?!.*[.][-'.]))[A-Za-z- '.]{2,64}$";
+
+    @Override
+    public void initialize(ValidName constraintAnnotation) {
+        // nothing to initialize
+    }
+
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/validators/NameValidator.java:18]:
```suggestion
```

Is this necessary?

File: src/main/java/app/coronawarn/quicktest/validation/validators/PhoneNumberValidator.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+public class PhoneNumberValidator implements ConstraintValidator<ValidPhoneNumber, String> {
+    private static final String PHONENUMBER_PATTERN = "^[0-9]*$";
+
+    @Override
+    public void initialize(ValidPhoneNumber constraintAnnotation) {
+        // nothing to initialize
+    }
+
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/validators/PhoneNumberValidator.java:17]:
```suggestion
```

Is this necessary?

File: src/main/java/app/coronawarn/quicktest/validation/validators/ShortHashValidator.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+public class ShortHashValidator implements ConstraintValidator<ValidShortHash, String> {
+    private static final String SHORTHASH_PATTERN = "^[A-Fa-f0-9]{8}$";
+
+    @Override
+    public void initialize(ValidShortHash constraintAnnotation) {
+        // nothing to initialize
+    }
+
```
Comment (f11h) [inline src/main/java/app/coronawarn/quicktest/validation/validators/ShortHashValidator.java:17]:
```suggestion
```

Is this necessary?

File: src/main/resources/application.yml [hunk 19]
```diff
@@ -1,8 +1,17 @@
...
+      enabled: true
+      path: '/h2-console'
   liquibase:
     change-log: classpath:/db/changelog.yml
   application:
     name: cwa-quick-test-backend
+  jpa:
+    generate-ddl: true
+    hibernate:
+      ddl-auto: update
```
Comment (f11h) [inline src/main/resources/application.yml:13]: This is not suitable for the master branch :D
```suggestion
    hibernate:
      ddl-auto: validate
```

File: src/main/resources/application.yml [hunk 20]
```diff
@@ -18,11 +27,25 @@ cwa-testresult-server:
   url: http://localhost:1234
 
 keycloak:
-  auth-server-url:
+  auth-server-url: https://quicktest-iam-6155b8d7d3c6.coronawarn.app/auth/
```
Comment (f11h) [inline src/main/resources/application.yml:30]: Please do not expose dev env urls

```suggestion
  auth-server-url:
```

File: src/main/resources/application.yml [hunk 21]
```diff
@@ -18,11 +27,25 @@ cwa-testresult-server:
...
   realm: c19quicktest
   resource: quick-test-portal
   public-client: true
   bearer-only: true
+  proxy-url: http://sia-lb.telekom.de:8080
```
Comment (f11h) [inline src/main/resources/application.yml:35]:
```suggestion
```

File: src/main/resources/db/changelog.yml [hunk 22]
```diff
@@ -1,4 +1,7 @@
 databaseChangeLog:
   - include:
       file: changelog/create-tables.yml
+      relativeToChangelogFile: true
+  - include:
+      file: changelog/testdata.yml
       relativeToChangelogFile: true
```
Comment (f11h) [inline src/main/resources/db/changelog.yml:7]:
```suggestion
      relativeToChangelogFile: true
      
```

File: src/main/resources/db/changelog/testdata.yml [hunk 23]
```diff
@@ -0,0 +1,44 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-testdata
```
Comment (f11h) [inline src/main/resources/db/changelog/testdata.yml:3]:
```suggestion
  - changeSet:
      id: insert-testdata
      context: test
```

File: src/main/resources/application.yml [hunk 24]
```diff
@@ -1,8 +1,17 @@
 spring:
+  h2:
+    console:
+      enabled: true
+      path: '/h2-console'
   liquibase:
     change-log: classpath:/db/changelog.yml
```
Comment (f11h) [inline src/main/resources/application.yml:7]:
```suggestion
    change-log: classpath:/db/changelog.yml
    contexts: prod
```

Add liquibase context to not add test data on real environments.
We should discuss whether we will put this on in application.yml or application-cloud.yml. Maybe the testdata is usefull for local execution.

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+        quickTest.setUpdatedAt(LocalDateTime.of(2021, 4, 8, 8, 11, 12));
+        quickTest.setFirstName("Joe");
+        quickTest.setLastName("Miller");
+        quickTest.setStreet("Boe");
+        quickTest.setHouseNumber("11");
+        quickTest.setInsuranceBillStatus(Boolean.FALSE);
+        quickTest.setSex(Sex.DIVERSE);
+        assertEquals("QuickTest(shortHashedGuid=cjfybkfn, hashedGuid=mkamhvdumyvhxeftazravmyrasozuloaghgluvbfjohpofogkylcnsybubamwnht, createdAt=2021-04-08T08:11:11, updatedAt=2021-04-08T08:11:12, confirmationCwa=true, tenantId=4711, testSpotId=4711-A, testResult=5, version=null, insuranceBillStatus=false, lastName=Miller, firstName=Joe, email=test@test.test, phoneNumber=00491777777777777, sex=DIVERSE, street=Boe, houseNumber=11, zipCode=12345, city=oyvkpigcga, testBrandId=AT116/21, testBrandName=Panbio (TM) Covid-19 Ag Rapid Test Device (Nasal))",quickTest.toString());
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:60]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+    void checkShortHashedGuidSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setShortHashedGuid(generators.generateString(13));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:65]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+    void checkHashedGuidSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setHashedGuid(generators.generateString(65));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:70]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+    void checkTenantIdSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setTenantId(generators.generateString(260));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:75]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+    void checkTestSpotIdSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setTestSpotId(generators.generateString(260));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:80]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+    void checkTestResultSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setTestResult(Short.parseShort("1"));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:85]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+    void checkTestResultDefault() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        Method method = QuickTest.class.getDeclaredMethod("onCreate");
+        method.setAccessible(true);
+        QuickTest quickTest = new QuickTest();
+        method.invoke(quickTest);
+        assertEquals(Short.parseShort("5"),quickTest.getTestResult());
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:93]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+        method.setAccessible(true);
+        QuickTest quickTest = new QuickTest();
+        method.invoke(quickTest);
+        LocalDateTime localDateTime = LocalDateTime.now();
+        assertTrue(quickTest.getCreatedAt() instanceof LocalDateTime);
+        log.info(localDateTime.toString());
+        log.info(quickTest.getCreatedAt().toString());
+        assertTrue(localDateTime.isEqual(quickTest.getCreatedAt()) || localDateTime.isAfter(quickTest.getCreatedAt()));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:105]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+        method.setAccessible(true);
+        QuickTest quickTest = new QuickTest();
+        method.invoke(quickTest);
+        LocalDateTime localDateTime = LocalDateTime.now();
+        assertTrue(quickTest.getUpdatedAt() instanceof LocalDateTime);
+        log.info(localDateTime.toString());
+        log.info(quickTest.getUpdatedAt().toString());
+        assertTrue(localDateTime.isEqual(quickTest.getUpdatedAt()) || localDateTime.isAfter(quickTest.getUpdatedAt()));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:117]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 34]
```diff
@@ -0,0 +1,151 @@
...
+    void checkVersionNotHasSetter() {
+        try {
+            Method method = QuickTest.class.getDeclaredMethod("setVersion");
+            assertTrue(false, "setVersion method not allowed");
+        } catch (NoSuchMethodException e) {
+            assertTrue(true);
+        }
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:126]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 35]
```diff
@@ -0,0 +1,151 @@
...
+    void checkEmailValidator() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setEmail(generators.generateString(13));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:131]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 36]
```diff
@@ -0,0 +1,151 @@
...
+    void checkPhoneNumberSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setPhoneNumber(generators.generateString(130));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:136]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 37]
```diff
@@ -0,0 +1,151 @@
...
+    void checkZipCodeSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setZipCode(generators.generateString(12));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:141]:
```suggestion

    @Test
```

File: src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java [hunk 38]
```diff
@@ -0,0 +1,151 @@
...
+    void checkTestBrandIdSize() {
+        QuickTest quickTest = new QuickTest();
+        quickTest.setTestBrandId(generators.generateString(13));
+    }
+    @Test
```
Comment (f11h) [inline src/test/java/app/coronawarn/quicktest/entity/QuickTestTest.java:146]:
```suggestion

    @Test
```

File: src/test/resources/application.yml [hunk 39]
```diff
@@ -3,6 +3,7 @@ spring:
     active: test
   liquibase:
     change-log: classpath:/db/changelog.yml
+    contexts: test
```
Comment (f11h) [inline src/test/resources/application.yml:6]: Added test context for test data creation in test execution

**Comment by f11h** [comment]
@mbaude @bjarne1337 @agaricusbisporus Can you please review my comments - It's mostly just about some little things

File: src/main/resources/application.yml [hunk 40]
```diff
@@ -1,8 +1,17 @@
 spring:
+  h2:
+    console:
+      enabled: true
+      path: '/h2-console'
```
Comment (f11h) [inline src/main/resources/application.yml:5]:
```suggestion
```

**Review by f11h — COMMENTED** [review]
thx for fixing :)
Just one question regarding the "empty todos" is open.

**Review by bjarne1337 — APPROVED** [review]

**Pull request merged**